src/ParserCombinators/Datatypes.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE GADTSyntax #-}
+{-# LANGUAGE InstanceSigs #-}
+-- {-# LANGUAGE FlexibleInstances #-}
+-- {-# LANGUAGE UndecidableInstances #-}
+
+module ParserCombinators.Datatypes where
+
+import Control.Applicative (Alternative (..))
+import Control.Monad (MonadPlus)
+
+------------------------
+---- ParseErrorType ----
+------------------------
+
+data ParseErrorType i e
+  = EndOfInput
+  | Unexpected i
+  | Expected i i
+  | ExpectedEndOfFile i
+  | CustomError e
+  | Empty
+  deriving (Eq)
+
+instance (Show i, Show e) => Show (ParseErrorType i e) where
+  show :: ParseErrorType i e -> String
+  show EndOfInput = "End of Input"
+  show (Unexpected a) = "Unexpected " ++ show a
+  show (Expected e a) = "Expected " ++ show e ++ ", got " ++ show a
+  show (ExpectedEndOfFile a) = "Expected EndOfFile, got " ++ show a
+  show (CustomError e) = "Custom Error: " ++ show e
+  show Empty = "No error"
+
+--------------------
+---- ParseError ----
+--------------------
+
+type Offset = Integer
```

Comment:
Probably use a newtype for this

Reply:
Any reasons to why that's better? I currently use the `succ`-ability (`Enum`) and `show`-ability (`Show`) of the type directly.

Reply:
Those are easy(one line) to recover with GND.

On the other hand, you can easily mess this up if you ever have another `Integer` in the same function and accidentally swap their names or pass them in the wrong order (and so on)

Reply:
What is GND?